### DISCUSSION TITLE
IConnection lifetime management

### DISCUSSION THREAD

**Discussion by ninjafinne** [Kistler-Group/sdbus-cpp#184] [Ideas]
The createConnection() family of functions return a std::unique_ptr\<sdbus::IConnection\>.

It's then possible to share this connection with multiple proxies by dereferencing and passing by reference.
A shared_ptr would more clearly show the intent.

Am I missing something?

**Comment by dleeds-cpi** [comment]
You can convert a `unique_ptr` to a `shared_ptr` if necessary. See https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr

**Reply by ninjafinne** [reply]
Yes, the intent however is more clear with a shared_ptr and the code becomes easier to read.

**Reply by sangelovic** [reply]
Correct, a `shared_ptr` would clearly show the intent *if* we would want to have a concept of a connection whose ownership (and thus lifetime) control is shared among multiple proxies.

That use case was not intended in the original design, however. The idea was that there is one owner of the connection, which may very often be the main application class (quite typical in case the application is a D-Bus service itself). This class may typically handle the event loop upon the connection, in addition to potentially other sources of events. In this case, such a connection lives during the entire app lifetime, and proxies (which may come and go) re-use this connection. In such application design, the lifespan of proxies (should) never exceed the lifespan of the connection. And the connection is not truly shared. It's merely "observed" and used by proxies. That observation is idiomatically expressed through a reference.

**Reply by ninjafinne** [reply]
I understand what I was missing. Thank you for thoroughly explaining.
